orttraining/orttraining/training_ops/cuda/collective/nccl_kernels.cc

```diff
@@ -43,78 +53,201 @@ Status NcclAllGather::ComputeInternal(OpKernelContext* context) const {
+          offset += tensor_bytes;
+        }
+      }
+    }
   }
+  return Status::OK();
+}
 
-  // AllGather.
-  const void* fusion_data_rank_offset = (const int8_t*)fusion_data + rank_start;
-  NCCL_RETURN_IF_ERROR(ncclAllGather(fusion_data_rank_offset, fusion_data, rank_count, dtype, comm, stream));
+NcclReduce::NcclReduce(const OpKernelInfo& info) : NcclKernel(info) {
+  ORT_ENFORCE(info.GetAttr<int64_t>("root_rank", &root_rank_).IsOK());
+  info.GetAttrOrDefault<int64_t>("num_input_readies", &num_input_readies_, static_cast<int64_t>(0));
+}
 
-  // Copy AllGather results to outputs.
-  offset = 0;
-  for (int i = 0; i < context->InputCount(); i++) {
+Status NcclReduce::ComputeInternal(OpKernelContext* context) const {
+  ORT_ENFORCE(context->InputCount() > 0);
+  auto onnx_type = context->Input<Tensor>(0)->DataType();
+  const size_t element_size = onnx_type->Size();
+
+  // Count total number of elements to Reduce.
+  int64_t total_count = 0;
+  for (int i = 0; i < context->InputCount() - num_input_readies_; i++) {
+    const Tensor* input_tensor = context->Input<Tensor>(i);
+    total_count += input_tensor->Shape().Size();
+  }
+
+  //When the contiguous memory is enabled, can remove this buffer
+  const int size = total_count * element_size;
+  const int64_t alignment = 32;
+  const int64_t padded_size = size + alignment - (size % alignment);
+  auto fusion_buffer = GetScratchBuffer<void>(padded_size);
+  void* fusion_data = fusion_buffer.get();
+
+  // Copy inputs to fusion buffer.
+  int64_t offset = 0;
+  for (int i = 0; i < context->InputCount() - num_input_readies_; i++) {
     const Tensor* input_tensor = context->Input<Tensor>(i);
-    const TensorShape& input_shape = input_tensor->Shape();
     const int64_t tensor_bytes = input_tensor->SizeInBytes();
-    Tensor* output_tensor = context->Output(i, input_shape);
 
-    // TODO: temporary hack until View is improved (it doesn't work with Alias)
-    output_tensor->SetByteOffset(input_tensor->ByteOffset());
+    void* fusion_data_at_offset = (int8_t*)fusion_data + offset;
+    const void* input_data = input_tensor->DataRaw();
+    CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(fusion_data_at_offset, input_data, tensor_bytes, cudaMemcpyDeviceToDevice));
 
-    // Only copy outputs that came from other ranks.
-    if (offset < rank_start || offset >= rank_end) {
-      void* output_data = output_tensor->MutableDataRaw();
+    offset += tensor_bytes;
+  }
+
+  const int rank = nccl_->Rank(group_type_);
+  cudaStream_t stream = nullptr;  //Default stream
+  ncclComm_t comm = nccl_->Comm(group_type_);
+  ncclDataType_t dtype = GetNcclDataType(onnx_type);
+  if (rank == root_rank_)
+    NCCL_RETURN_IF_ERROR(ncclReduce(fusion_data, fusion_data, total_count, dtype, ncclSum, root_rank_, comm, stream));
+  else
+    NCCL_RETURN_IF_ERROR(ncclReduce(fusion_data, NULL, total_count, dtype, ncclSum, root_rank_, comm, stream));
```

Comment:
(nit):  you don't need this `if` condition.  `ncclReduce()` will already ignore the output buffer for ranks that aren't the root_rank.  You can just do:

```
NCCL_RETURN_IF_ERROR(ncclReduce(fusion_data, fusion_data, total_count, dtype, ncclSum, root_rank_, comm, stream));
```

Reply:
Got it, will do.